packages/core/src/layout/Sidebar/Items.tsx

```diff
@@ -114,100 +123,103 @@ const useStyles = makeStyles<BackstageTheme>(theme => {
-      >
-        <Icon fontSize="small" className={classes.icon} />
-      </Badge>
-    );
+type SidebarItemLinkProps = SidebarItemBaseProps & {
+  text?: string;
+  to: string;
+  onClick?: (ev: React.MouseEvent) => void;
+};
 
-    const childProps = {
-      onClick,
-      className: clsx(classes.root, isOpen ? classes.open : classes.closed),
-    };
+type SidebarItemProps = SidebarItemButtonProps | SidebarItemLinkProps;
 
-    if (!isOpen) {
-      if (to === undefined) {
-        return (
-          <div {...childProps} ref={ref}>
-            {itemIcon}
-          </div>
-        );
-      }
+function isButtonItem(
+  props: SidebarItemProps,
+): props is SidebarItemButtonProps {
+  return (props as SidebarItemLinkProps).to === undefined;
+}
 
-      return (
-        <NavLink
-          {...childProps}
-          activeClassName={classes.selected}
-          to={to}
-          end
-          ref={ref}
-        >
-          {itemIcon}
-        </NavLink>
-      );
-    }
+export const SidebarItem = forwardRef<any, SidebarItemProps>((props, ref) => {
+  const {
+    icon: Icon,
+    text,
+    hasNotifications = false,
+    onClick,
+    children,
+  } = props;
+  const classes = useStyles();
+  // XXX (@koroeskohr): unsure this is optimal. But I just really didn't want to have the item component
+  // depend on the current location, and at least have it being optionally forced to selected.
+  // Still waiting on a Q answered to fine tune the implementation
+  const { isOpen } = useContext(SidebarContext);
 
-    const content = (
-      <>
-        <div data-testid="login-button" className={classes.iconContainer}>
-          {itemIcon}
-        </div>
-        {text && (
-          <Typography variant="subtitle2" className={classes.label}>
-            {text}
-          </Typography>
-        )}
-        <div className={classes.secondaryAction}>{children}</div>
-      </>
-    );
+  const itemIcon = (
+    <Badge
+      color="secondary"
+      variant="dot"
+      overlap="circle"
+      invisible={!hasNotifications}
+    >
+      <Icon fontSize="small" className={classes.icon} />
+    </Badge>
+  );
 
-    if (to === undefined) {
-      return (
-        <div {...childProps} ref={ref}>
-          {content}
-        </div>
-      );
-    }
+  const closedContent = itemIcon;
+
+  const openContent = (
+    <>
+      <div data-testid="login-button" className={classes.iconContainer}>
+        {itemIcon}
+      </div>
+      {text && (
+        <Typography variant="subtitle2" className={classes.label}>
+          {text}
+        </Typography>
+      )}
+      <div className={classes.secondaryAction}>{children}</div>
+    </>
+  );
+
+  const content = isOpen ? openContent : closedContent;
 
+  const childProps = {
+    onClick,
+    className: clsx(
+      classes.root,
+      isOpen ? classes.open : classes.closed,
+      isButtonItem(props) && classes.buttonItem,
+    ),
+  };
+
+  if (isButtonItem(props)) {
```

Comment:
Think this could be converted to `const isButton = (props as SidebarItemLinkProps).to === undefined` further up, for reuse above and to avoid the extra function

Reply:
Hi @Rugvip, thank you for reviewing the pr! So I started with that initially, but looks like with type-guards to properly type narrow each branch, they need to be inline in the branch section, either as an expression or type-guard function. So as a constant above, the `props.to` complains:

<img width="519" alt="image" src="https://user-images.githubusercontent.com/8230000/103308708-edb03600-49c7-11eb-9144-af392ce8d104.png">

I tried moving that expression directly into the if statement, but that actually doesn't work either, since it's not a valid inline type-guard. It seems like the closest valid guard would be something like: `"to" in props`: 
<img width="519" alt="image" src="https://user-images.githubusercontent.com/8230000/103309169-35838d00-49c9-11eb-8743-80a1ca80cd43.png">

which type-checks, but weirdly, actually breaks the Sidebar.stories.tsx here: https://github.com/backstage/backstage/blob/0acf4c92992c7a243d6b9a44d328f4c91ad4b8c7/packages/core/src/layout/Sidebar/Items.tsx#L250
I guess it's a slightly different check than `props.to === undefined`? Given all this, I ultimately settled on the user-defined type guard, to keep the runtime logic the same and get the type-checking.

If all of this is unnecessarily complicated just to to have the type union, happy to just make `to` an optional prop. lmk!